runners/prism/java/src/main/java/org/apache/beam/runners/prism/PrismRunner.java

```diff
@@ -72,15 +72,50 @@ public PipelineResult run(Pipeline pipeline) {
         prismPipelineOptions.getDefaultEnvironmentType(),
         prismPipelineOptions.getJobEndpoint());
 
-    return internal.run(pipeline);
+    Runnable closer = runPrismAndProvideCloser(prismPipelineOptions);
+    try {
+      JobServicePipelineResult delegateResult = (JobServicePipelineResult) internal.run(pipeline);
+      PrismPipelineResult result = new PrismPipelineResult(delegateResult, closer);
+      CompletableFuture<?> ignored =
+          delegateResult.getTerminalStateFuture().whenComplete(result::onJobStateComplete);
+      if (isInvokedInTest()) {
+        LOG.info("invoking Pipeline::waitUntilFinish due to invoking in a class named ^.*Test$");
+        result.waitUntilFinish();
+      }
+      return result;
+    } catch (RuntimeException e) {
+      closer.run();
+      throw e;
+    }
+  }
+
+  private boolean isInvokedInTest() {
+    for (StackTraceElement element : Thread.currentThread().getStackTrace()) {
+      if (element.getClassName().endsWith("Test")) {
+        return true;
+      }
+    }
+    return false;
   }
 
-  private static void assignDefaultsIfNeeded(PrismPipelineOptions prismPipelineOptions) {
-    if (Strings.isNullOrEmpty(prismPipelineOptions.getDefaultEnvironmentType())) {
-      prismPipelineOptions.setDefaultEnvironmentType(Environments.ENVIRONMENT_LOOPBACK);
+  private static void assignDefaultsIfNeeded(PrismPipelineOptions options) {
+    if (Strings.isNullOrEmpty(options.getDefaultEnvironmentType())) {
+      options.setDefaultEnvironmentType(Environments.ENVIRONMENT_LOOPBACK);
     }
-    if (Strings.isNullOrEmpty(prismPipelineOptions.getJobEndpoint())) {
-      prismPipelineOptions.setJobEndpoint(DEFAULT_PRISM_ENDPOINT);
+    if (Strings.isNullOrEmpty(options.getJobEndpoint())) {
+      options.setJobEndpoint(DEFAULT_PRISM_ENDPOINT);
```

Comment:
The default behavior shouldn't be a port that prism selects, but the one this caller selects *for* prism. There's no guarantee that 8073 is ever free or remains free.

Reply:
That's an easy fix. I can have it select randomly based on the freely available.